apps/rush-lib/src/logic/buildCache/ProjectBuildCache.ts

```diff
@@ -194,61 +229,103 @@ export class ProjectBuildCache {
     }
 
     terminal.writeVerboseLine(`Caching build output folders: ${filteredOutputFolders.join(', ')}`);
-    let encounteredTarErrors: boolean = false;
-    const tarStream: stream.Readable = tar.create(
-      {
-        gzip: true,
-        portable: true,
-        strict: true,
-        cwd: projectFolderPath,
-        filter: (tarPath: string, stat: tar.FileStat) => {
-          const tempStats: fs.Stats = new fs.Stats();
-          tempStats.mode = stat.mode;
-          if (tempStats.isSymbolicLink()) {
-            terminal.writeError(`Unable to include "${tarPath}" in build cache. It is a symbolic link.`);
-            encounteredTarErrors = true;
-            return false;
-          } else {
-            return true;
-          }
+    const tarExecutablePath: string | undefined = ProjectBuildCache._tarExecutablePath;
+    let localCacheEntryPath: string | undefined;
+
+    if (tarExecutablePath) {
+      const tempLocalCacheEntryPath: string = this._localBuildCacheProvider.getCacheEntryPath(cacheId);
+      const childProcess: ChildProcess = Executable.spawn(
+        tarExecutablePath,
+        ['-c', '-f', tempLocalCacheEntryPath, '-z', ...filteredOutputFolders],
+        {
+          currentWorkingDirectory: projectFolderPath
         }
-      },
-      filteredOutputFolders
-    );
-    const cacheEntryBuffer: Buffer = await this._readStreamToBufferAsync(tarStream);
-    if (encounteredTarErrors) {
-      return false;
+      );
+      const writeLocalCacheSuccess: boolean = await new Promise((resolve: (result: boolean) => void) => {
+        childProcess.on('exit', (code) => {
+          resolve(code === 0);
+        });
+      });
+
+      if (writeLocalCacheSuccess) {
+        localCacheEntryPath = tempLocalCacheEntryPath;
+      }
     }
 
-    const setLocalCacheEntryPromise: Promise<boolean> = this._localBuildCacheProvider.trySetCacheEntryBufferAsync(
-      terminal,
-      cacheId,
-      cacheEntryBuffer
-    );
+    let cacheEntryBuffer: Buffer | undefined;
+    let setLocalCacheEntryPromise: Promise<string> | undefined;
+    if (!localCacheEntryPath) {
+      // If we weren't able to create the cache entry with tar, try to do it with the "tar" NPM package
+      let encounteredTarErrors: boolean = false;
+      const tarStream: stream.Readable = tar.create(
+        {
+          gzip: true,
+          portable: true,
+          strict: true,
+          cwd: projectFolderPath,
+          filter: (tarPath: string, stat: tar.FileStat) => {
+            const tempStats: fs.Stats = new fs.Stats();
+            tempStats.mode = stat.mode;
+            if (tempStats.isSymbolicLink()) {
+              terminal.writeError(`Unable to include "${tarPath}" in build cache. It is a symbolic link.`);
```

Comment:
When the `tar` binary is executed (instead of using NPM tar), how is this "no symlinks" check applied?

Reply:
It's not unfortunately :/. I fear reading every file and folder in the cache folders to check if any are symlinks would cancel out any performance improvements.

Reply:
You can use `--dereference` to follow them and include the targets, but simply excluding I'm not sure if is available

Reply:
That argument isn't available on the copy of `tar` that comes with Windows.

Reply:
Would it be inefficient for JavaScript to build the list of paths to be tarred, and then the `tar` binary only does the reading+compressing part?

The Windows `tar` command is so feature limited, I wonder if we'll really be able to use it directly.  Today you don't do any filtering of paths, but that seems very likely to change as we build out more features to solve whatever problems come along.

Reply:
These safeguards are important -- traditionally the user experience tradeoff goes like:
- **The cache is a bit slow** --> *"I guess I'll get some coffee while this build finishes."*
- **The build failed because of a cache issue.** --> *"Ugh. Did I make a simple mistake in my configuration? Or is it a thorny bug in the cache algorithm? No time to investigate that. EVERYBODY, THE FIX IS TO DISABLE CACHING!"*

Reply:
Resolved this by enumerating paths explicitly